README
credsman version 1.00
=====================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2019 by A. U. Thor

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.30.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension credsman.

0.01  Sun Oct  6 01:25:37 2019
	- original version; created by h2xs 1.23 with options
		-A -n credsman



exmaple.pl
    use strict;
    use warnings;
    use ExtUtils::testlib;
    use credsman qw(login);

    # This type of function is necessary to run login, 
    # you need to handle the access or conenction and Error messages

    sub Connect_Example {
        my $credentials = shift;
        # Here your code to login or connect using user and password
        if( $credentials->{user} eq 'pepe' and  $credentials->{password} eq 'pepepass' ){
            print "The Target Name is: $credentials->{target}\n";
            print "  User  : $credentials->{user}\n";
            print "  Pass  : $credentials->{password}\n";
            print "Attempt : $credentials->{attempt} of $credentials->{limit}\n"; 
            # Return 0 - Success
            return 0;
        }
        else{
            print "Fail\n";
            # Return to fail
            return 1;
        }
    }
   
    # In this Example the prgram will die at the attempt number 10.

    die "No Zero Return" if login( 
        program  => 'credsman',         # The Prefix to Store the credentials in wcm 
        target   => "Test",             # The Target to validate user and password, usually a server
        subref   => \Connect_Example(),  # Reference to a Function (how to validate password)
        limit    => 10,                 # Number of Attemps before the program Finish
    );

Makefile.PL
use 5.018000;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME               => 'credsman',
    VERSION_FROM       => 'lib/credsman.pm', # finds $VERSION, requires EU::MM from perl >= 5.5
    CONFIGURE_REQUIRES => { "ExtUtils::MakeMaker" => 0, },
    PREREQ_PM          => {'Test::More' => 0, 
                           'ExtUtils::MakeMaker' => '7.12',
                           'Types::Standard' => 0,
                           'Params::ValidationCompiler' => 0,
                           'Data::Dumper' => 0,
                           'Exporter'     => 0,
                          }, 
    ABSTRACT_FROM      => 'lib/credsman.pm', # retrieve abstract from module
    AUTHOR             => 'Rodrigo Agurto <rodagurto@aim.com>',
    LICENSE            => 'perl',
    #Value must be from legacy list of licenses here
    #http://search.cpan.org/perldoc?Module%3A%3ABuild%3A%3AAPI
    LIBS               => ['-lgdi32 -lCredui'], # e.g., '-lm'
    DEFINE             => '', # e.g., '-DHAVE_SOMETHING'
    OPTIMIZE           => '-O3',
    TEST_REQUIRES      => {
                           'Test::More' => 0,
                           'ExtUtils::testlib' => 0,
                           'Params::ValidationCompiler' => 0,
                        },
);


lib/credsman.pm
package credsman;
use 5.018000;
use strict;
use warnings;
use Types::Standard qw[Int Str CodeRef];
use Params::ValidationCompiler qw[validation_for];
use Data::Dumper;
use Exporter qw(import);
our @EXPORT_OK = qw[login GuiCred];

our $VERSION = '1.00';

require XSLoader;
XSLoader::load('credsman', $VERSION);

# There are 4 XS internal functions that interact with Credential Manager
# RemoveCredentials  - Remove Credentials
# SaveCredentials    - Store Credentials
# GuiCredentials     - Open Prompt USER and PASSWORD gui 
# GuiCred            - Same of GuiCredentials but exposed to this module
# work_name          - Creates Credentials Name String 

#-------------------------------------------------------------------------------------------#
my $validator = validation_for(
    params => {
        program   => { type => Str },
        target    => { type => Str },
        subref    => { type => CodeRef },
        limit     => { type => Int, optional => 1, default => 3 },
        debug     => { type => Int, optional => 1, default => 0 }
    }
);
#-------------------------------------------------------------------------------------------#
sub login{
    my %arg = $validator->(@_);
    say  "*** Arguments ****\n".Dumper \%arg if $arg{debug};
    my %wrkCred = (
        status   => 5,
        attempt  => 0,
        limit    => $arg{limit},
        password => undef,
        user     => undef,
        target   => $arg{target},
    );
    # Concat Target Name - This is the name to be stored 
    my $TargetName = work_name($arg{program},$arg{target});
    say "TargetName : ".$TargetName if $arg{debug};
    # Load Passwords ad runs the function passed with the argument
    while ($wrkCred{status} != 0 and $wrkCred{attempt} < $wrkCred{limit}){
        say "in loop" if $arg{debug};
        # load Credentials from Windows Credential Manager
        ($wrkCred{user}, $wrkCred{password}) = @{LoadCredentials($TargetName)};
        say "*** Load Credentials ****\n".Dumper \%wrkCred if $arg{debug};
        # Check if the Credentials was found
        if( !defined $wrkCred{user}){
            # Request User and Password GUI
            say "Not Defined" if $arg{debug};
            ($wrkCred{user}, $wrkCred{password}) = @{
                GuiCredentials(
                    $arg{program}, 
                    $arg{target}, 
                    $TargetName, 
                    $wrkCred{attempt})
            };
            # No user or password passed, or cancel
            unless(defined $wrkCred{user} and defined $wrkCred{password}){
                die "Cancel" ;
            }

            say "After GUI \n".Dumper \%wrkCred if $arg{debug};
            # Store Credentials in Windoes Credential Manager
            if( SaveCredentials($TargetName, $wrkCred{user}, $wrkCred{password}) ){
                die "Error to Save Credentials";
            };
        }
        else {
            # Call Function passed in Argument
            $wrkCred{status} = $arg{subref}->(\%wrkCred);
            # Expeciting RC 0 to pass
            if ($wrkCred{status}){
                $wrkCred{attempt}++;
                RemoveCredentials($TargetName);
            }
        }
    }
    # Clear Credentials for Security
    $wrkCred{user}     =~ s/.*/ /g;
    $wrkCred{password} =~ s/.*/ /g;
    say "END" if $arg{debug};
    # Return Status
    return $wrkCred{status};
}
#-------------------------------------------------------------------------------------------#
my $GuiVal = validation_for(
    params => {
        message   => { type => Str },
        caption   => { type => Str },
        attempt   => { type => Int, optional => 1, default => 0 }
    }
);
#-------------------------------------------------------------------------------------------#
sub GuiCred {
    my %arg = $GuiVal->(@_);
    return GuiCredentials(
            $arg{caption}, 
            $arg{message}, 
            '', 
            $arg{attempt});
}

1;
__END__

=head1 NAME

credsman - is a simple Pel extension to work with 'Windows Credential Manager'.  

=head1 SYNOPSIS

    use strict;
    use warnings;
    use ExtUtils::testlib;
    use credsman qw(login);

    # This type of function is necessary to run login, 
    # you need to handle the access or conenction and Error messages

    sub Connect_Example {
        my $credentials = shift;
        # Here your code to login or connect using user and password
        if( $credentials->{user} eq 'pepe' and  $credentials->{password} eq 'pepepass' ){
            print "The Target Name is: $credentials->{target}\n";
            print "  User  : $credentials->{user}\n";
            print "  Pass  : $credentials->{password}\n";
            print "Attempt : $credentials->{attempt} of $credentials->{limit}\n"; 
            # Return 0 - Success
            return 0;
        }
        else{
            print "Fail\n";
            # Return to fail
            return 1;
        }
    }
   
    # In this Example the prgram will die at the attempt number 10.

    die "No Zero Return" if login( 
        program  => 'credsman',          # The Prefix to Store the credentials in wcm 
        target   => "Test",              # The Target to validate user and password, usually a server
        subref   => \Connect_Example(),  # Reference to a Function (how to validate password)
        limit    => 10,                  # Number of Attemps before the program Finish
    );


=head1 DESCRIPTION

Stub documentation for credsman, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

    # Windows Credential Manager - Generic Credentials
    # format:
    # *['program name']~['Server name or Addres']*


Blah blah blah.

=head2 EXPORT

None by default.



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

A. U. Thor, E<lt>a.u.thor@a.galaxy.far.far.awayE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2019 by A. U. Thor

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.30.0 or,
at your option, any later version of Perl 5 you may have available.


=cut
